libraries/AP_BattMonitor/AP_BattMonitor_AD7091R5.cpp

```diff
@@ -0,0 +1,266 @@
+ *
+ * @param mon
+ * @param mon_state
+ * @param params
+ */
+AP_BattMonitor_AD7091R5::AP_BattMonitor_AD7091R5(AP_BattMonitor &mon,
+        AP_BattMonitor::BattMonitor_State &mon_state,
+        AP_BattMonitor_Params &params) :
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+/**
+ * @brief probe and initialize the sensor and register call back
+ *
+ */
+void AP_BattMonitor_AD7091R5::init()
+{
+
+    _state.healthy = false;
+    // voltage and current pins from params and check if there are in range
+    if (_volt_pin.get() >= AD7091R5_BASE_PIN && _volt_pin.get() <= AD7091R5_BASE_PIN + AD7091R5_NO_OF_CHANNELS) {
+        volt_buff_pt = _volt_pin.get() - AD7091R5_BASE_PIN;
+
+        if (_curr_pin.get() >= AD7091R5_BASE_PIN && _curr_pin.get() <= AD7091R5_BASE_PIN + AD7091R5_NO_OF_CHANNELS) {
+            curr_buff_pt = _curr_pin.get() - AD7091R5_BASE_PIN;
+            _state.healthy = true; //set status healthy
+        }
+    }
+
+    //return if configured incorrectly
+    if (!_state.healthy) {
+        return;
+    }
+
+    // only the first instance read the i2c device
+    if (_first) {
+        _first = false;
+        _state.healthy = false;
+        // probe i2c device
+        _dev = hal.i2c_mgr->get_device(AD7091R5_I2C_BUS, AD7091R5_I2C_ADDR);
+
+        if (_dev) {
+            WITH_SEMAPHORE(_dev->get_semaphore());
+            _dev->set_retries(10); // lots of retries during probe
+            //Reset and config device
+            if (_reset() && _config()) {
+                _state.healthy = true;
+                _dev->set_retries(2); // drop to 2 retries for runtime
+                _dev->register_periodic_callback(AD7091R5_PERIOD_USEC, FUNCTOR_BIND_MEMBER(&AP_BattMonitor_AD7091R5::_read_adc, void));
+            }
+        }
+    }
+}
+
+/**
+ * @brief read - read the voltage and curren
+ *
+ */
+void AP_BattMonitor_AD7091R5::read()
+{
+
+    WITH_SEMAPHORE(sem);
+    //copy global health status to all instances
+    _state.healthy = _health;
+
+    //return if system not healthy
+    if (!_state.healthy) {
+        return;
+    }
+
+    //voltage conversion
+    _state.voltage = (_data_to_volt(_analog_data[volt_buff_pt].data) - _volt_offset) * _volt_multiplier;
+
+    //current amps conversion
+    _state.current_amps = (_data_to_volt(_analog_data[curr_buff_pt].data) - _curr_amp_offset) * _curr_amp_per_volt;
+
+    // calculate time since last current read
+    uint32_t tnow = AP_HAL::micros();
+    uint32_t dt_us = tnow - _state.last_time_micros;
+
+    // update total current drawn since startup
+    update_consumed(_state, dt_us);
+
+    // record time
+    _state.last_time_micros = tnow;
+}
+
+/**
+ * @brief read all four channels and store the results
+ *
+ */
+void AP_BattMonitor_AD7091R5::_read_adc()
+{
+    uint8_t data[AD7091R5_NO_OF_CHANNELS*2] = {0};
```

Comment:
You shouldn't need the initializer, as it should be filled in by the transfer, and otherwise ignored.